20597 - Make BackoffSupervisor respect OneForOneStrategy's maxNumRetries property

Sushisource opened the pull request:
Regards #20597 

I think I've managed to follow the contributing guidelines correctly. I had some issues running sbt (namely, it seems to like to blow the stack while running scalariform - win10,x64,latest scala/java) but I was able to run the test suite through IntelliJ at least.

Thanks for your reviews!

Sushisource committed:
Fix for 20597 - Make BackoffSupervisor respect OneForOneStrategy's maxNumRetries property

ktoso commented:
OK TO TEST

ktoso commented on akka-actor-tests/src/test/scala/akka/pattern/BackoffOnRestartSupervisorSpec.scala:
```diff
@@ -169,5 +169,23 @@ class BackoffOnRestartSupervisorSpec extends AkkaSpec with ImplicitSender {
+        probe.watch(supervisor)
+        // Have the child throw an exception 5 times, which is more than the max restarts allowed.
+        for (i ← 1 to 5) {
+          println(s"iteration $i")
```
remove println please :)

Sushisource replied:
Whoops! That's an embarrassing mistake.

ktoso commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -38,7 +38,15 @@ private class BackoffOnRestartSupervisor(
+          val nextRestartCount = restartCount + 1
+          if (strategy.maxNrOfRetries >= 0 && nextRestartCount > strategy.maxNrOfRetries) {
+            // If we've exceeded the maximum # of retries allowed by the Strategy, die.
+            log.debug(s"Terminating on restart # $nextRestartCount which exceeds max allowed restarts (${strategy.maxNrOfRetries})")
```
Use markers instead (to avoid allocations if debug mode is disabled):
`log.debug(s"Terminating on restart #{} which exceeds max allowed restarts ({})", nextRestartCount, strategy.maxNrOfRetries)`

ktoso commented:
LGTM from my side

Sushisource committed:
Fixes for review comments

Sushisource commented:
@ktoso Fixed your comments. Thanks!

patriknw commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -38,7 +38,15 @@ private class BackoffOnRestartSupervisor(
           val childRef = sender()
-          become(waitChildTerminatedBeforeBackoff(childRef) orElse handleBackoff)
+          val nextRestartCount = restartCount + 1
+          if (strategy.maxNrOfRetries >= 0 && nextRestartCount > strategy.maxNrOfRetries) {
```
not time involved?

Sushisource replied:
Sorry took me forever to get back here -- I assume you're referring to the "withinTimeRange" parameter to the OneForOneStrategy? I can definitely add that into this PR if desired.

patriknw replied:
That would be great if you added. I don't think this can be useful without taking reset time into account.

ktoso commented:
Hi @Sushisource, would you be able to address Patrik's comment? This is almost done :-)

Sushisource committed:
Added support for withinTimeRange property of OneForOneSupervisor to BackoffSupervisor

Sushisource commented:
@ktoso @patriknw Apologies again for this taking so long, been on vacation. I've added support for the time check - the one thing I wasn't 100% on is if the places I'm resetting the "lastResetTime" property I added all make sense. I think they do, but let me know.

patriknw commented on akka-actor-tests/src/test/scala/akka/pattern/BackoffOnRestartSupervisorSpec.scala:
```diff
@@ -186,5 +191,27 @@ class BackoffOnRestartSupervisorSpec extends AkkaSpec with ImplicitSender {
+        // AFter waiting longer than the time range, throw again, we should not be restarted.
+        Thread.sleep(320)
+        supervisor ! "THROW"
+        // Superviser should've terminated.
```
Something wrong here, I think.

The `withinTimeRange` should be like a reset of the restart counter, and it only makes sense to use it together with `maxNrOfRetries`.

Example `maxNrOfRetries = 5, withinTimeRange = 10.seconds`.
If you throw 6 times within 10 seconds it will stop.
If you throw 4 times within 10 seconds and then later throw 3 more times it will still restart.

You can think of it as the counter is reset to 0 every 10 seconds.

Look at the tests in https://github.com/akka/akka/blob/master/akka-actor-tests/src/test/scala/akka/actor/RestartStrategySpec.scala

patriknw replied:
@Sushisource ping, what's the status here?

Sushisource replied:
Ah, that makes sense, I just misinterpreted the meaning of that param - this way definitely makes more sense. Thanks

Sushisource committed:
Fix review comments to make withinTimeRange work as expected

johanandren commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -27,6 +27,14 @@ private class BackoffOnRestartSupervisor(
   import BackoffSupervisor._
+
+  if (strategy.withinTimeRange.isFinite()) {
+    // If the user has defined a time range for the maxNrOfRetries, we'll queue up a message
```
"we'll schedule a message"

johanandren commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -27,6 +27,14 @@ private class BackoffOnRestartSupervisor(
+    // to ourselves every time that range elapses, to reset the restart counter.
+    val finiteWithinTimeRange = strategy.withinTimeRange.asInstanceOf[FiniteDuration]
+    system.scheduler.schedule(finiteWithinTimeRange, finiteWithinTimeRange, self, ResetRestartCount(restartCount))
+  }
```
Hmm. So this will actually keep sending messages even though we are not at all restarting, wouldn't a solution that only does processing when there was a restart be better?

Sushisource replied:
Good point, I'll take a look at fixing that

Sushisource committed:
Fixed unnecessarily scheduling the message if the child hasn't restarted

johanandren commented on akka-actor/src/main/scala/akka/pattern/BackoffOnRestartSupervisor.scala:
```diff
@@ -45,6 +38,13 @@ private class BackoffOnRestartSupervisor(
+            // to ourselves every time that range elapses, to reset the restart counter. We hide it
+            // behind this conditional to avoid queuing the message unnecessarily
+            val finiteWithinTimeRange = strategy.withinTimeRange.asInstanceOf[FiniteDuration]
+            system.scheduler.schedule(finiteWithinTimeRange, finiteWithinTimeRange, self, ResetRestartCount(restartCount))
```
`scheduleOnce` rather, as it will then set `restartCount` to `0` and then trigger this again if needed

patriknw commented:
LGTM, after fixing `scheduleOnce`

Sushisource committed:
Change the restart message to be scheduled once

Sushisource commented:
Thanks for bearing with me even though it took quite a while! Let me know if I need to do anything else to prep this for merging.

patriknw reviewed (approved):
LGTM, thanks

patriknw commented:
PLS BUILD
